--- a/stdlib/os/__init__.pyi
+++ b/stdlib/os/__init__.pyi
@@ -27,14 +27,11 @@ from subprocess import Popen
 from typing import IO, Any, AnyStr, BinaryIO, Generic, NoReturn, Protocol, TypeVar, overload, runtime_checkable
 from typing_extensions import Final, Literal, Self, TypeAlias, final

-from . import path as _path
+from . import path

 if sys.version_info >= (3, 9):
     from types import GenericAlias

-# This unnecessary alias is to work around various errors
-path = _path
-
 _T = TypeVar("_T")
 _T1 = TypeVar("_T1")
 _T2 = TypeVar("_T2")
