--- a/stdlib/enum.pyi
+++ b/stdlib/enum.pyi
@@ -113,35 +113,6 @@ class EnumMeta(ABCMeta):
     def __len__(self) -> int: ...
     def __bool__(self) -> Literal[True]: ...
     def __dir__(self) -> list[str]: ...
-    # Simple value lookup
-    @overload
-    def __call__(cls: type[_EnumMemberT], value: Any, names: None = None) -> _EnumMemberT: ...
-    # Functional Enum API
-    if sys.version_info >= (3, 11):
-        @overload
-        def __call__(
-            cls,
-            value: str,
-            names: _EnumNames,
-            *,
-            module: str | None = None,
-            qualname: str | None = None,
-            type: type | None = None,
-            start: int = 1,
-            boundary: FlagBoundary | None = None,
-        ) -> type[Enum]: ...
-    else:
-        @overload
-        def __call__(
-            cls,
-            value: str,
-            names: _EnumNames,
-            *,
-            module: str | None = None,
-            qualname: str | None = None,
-            type: type | None = None,
-            start: int = 1,
-        ) -> type[Enum]: ...
     _member_names_: list[str]  # undocumented
     _member_map_: dict[str, Enum]  # undocumented
     _value2member_map_: dict[Any, Enum]  # undocumented
